pkg/frontend/querymiddleware/step_align.go

```diff
@@ -7,20 +7,76 @@ package querymiddleware
 
 import (
 	"context"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+	"github.com/grafana/mimir/pkg/util/validation"
 )
 
+type stepAlignMiddleware struct {
+	next     Handler
+	limits   Limits
+	resolver tenant.Resolver
```

Comment:
Most other middlewares use the global `tenant.TenantIDs()` methods. Unfortunately those functions rely on global state which makes this a huge pain to test. Thus I'm explicitly using a non-global resolver here. I want to get rid of the global state in `dskit/tenant` but that's outside the scope of this PR.